actions/lib/action.py

```diff
@@ -12,16 +12,45 @@ def __init__(self, config):
         super(TerraformBaseAction, self).__init__(config)
         self.terraform = Terraform()
 
-    def check_result(self, return_code, stdout, stderr):
+    def check_result(self, return_code, stdout, stderr, return_output=False,
+                     valid_return_codes=None):
         """Check the return code from the terraform action and return the output with
         the error message (if there is one)
         :param return_code: return code from the Terraform command that was run
         :param stdout: output from the Terraform command that was run
         :param stderr: error message (if any) from the Terraform command
+        :param return_output: whether to return the results of `terraform output` command
+        :param valid_return_codes: list of valid return codes
         :returns: success flag for the st2 action and output from the Terraform command
         """
-        output = stdout + "\n" + stderr
-        if return_code == 0:
-            return (True, output)
+        if valid_return_codes is None:
+            valid_return_codes = [0]
+        if return_output:
+            output = None
+            if return_code in valid_return_codes:
+                output = self.terraform.output()
         else:
-            return (False, output)
+            output = TerraformBaseAction.concat_std_output(stdout, stderr)
+
+        if return_code in valid_return_codes:
+            return True, output
+        else:
+            return False, output
+
+    @staticmethod
+    def concat_std_output(stdout, stderr):
+        """Combines stdout and stderr from terraform execution, one or the other
+        can be None
+        :param stdout: results from stdout
+        :param stderr: results from stderr
+        :returns: string concatenation
+        """
+        output = None
```

Comment:
I'm not sure this is super helpful, because now you don't know which portions of the output are stdout vs stderr.

Reply:
I agree completely, this is the default behavior today from this pack. Both stdout and stderr are concatenated together and returned as the result.

It's unusable, but I kept it here to keep backward compatibility for actions that I haven't touched. But an argument could be made that all actions should get the new treatment.